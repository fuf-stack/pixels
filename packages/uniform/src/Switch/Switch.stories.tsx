import type { Meta, StoryObj } from '@storybook/react-vite';

// Import icons for the stories
import { FaEnvelope, FaFlask, FaPhone, FaRocket } from 'react-icons/fa';

import { action } from 'storybook/actions';
import { userEvent, within } from 'storybook/test';

import { SubmitButton } from '@fuf-stack/uniform';
import { boolean, veto } from '@fuf-stack/veto';

import { Form } from '../Form';
import Switch from './Switch';

const meta: Meta<typeof Switch> = {
  title: 'uniform/Switch',
  component: Switch,
  decorators: [
    (Story, { parameters }) => {
      return (
        <Form
          className="min-w-60"
          onSubmit={action('onSubmit')}
          {...(parameters?.formProps || {})}
        >
          <Story />
          <div className="mt-4 flex justify-end">
            <SubmitButton />
          </div>
        </Form>
      );
    },
  ],
};

export default meta;
type Story = StoryObj<typeof Switch>;

export const Default: Story = {
  args: {
    name: 'switchField',
    label: '🐛 Enable bugs in production',
  },
};

export const WithInitialValue: Story = {
  parameters: {
    formProps: { initialValues: { switchField: true } },
  },
  args: {
    name: 'switchField',
    label: '☕ Coffee is a vegetable',
  },
};

export const Disabled: Story = {
  args: {
    name: 'switchField',
    label: '📚 Read documentation first',
    disabled: true,
  },
};

const requiredValidation = veto({
  switchField: boolean(),
});

export const Required: Story = {
  parameters: { formProps: { validation: requiredValidation } },
  args: {
    label: '🚀 Deploy on Friday',
    name: 'switchField',
  },
};

const validation = veto({
  switchField: boolean().refine((value: boolean) => {
    return !value;
  }, 'Please keep your sanity intact'),
});

export const Invalid: Story = {
  parameters: { formProps: { validation } },
  args: {
    label: '🤡 Use jQuery in 2024',
    name: 'switchField',
  },
  play: async ({ canvasElement }) => {
    const canvas = within(canvasElement);
    const input = canvas.getByTestId('switchfield');
    await userEvent.click(input, {
      delay: 300,
    });
  },
};

export const WithThumbIcon: Story = {
  parameters: {
    formProps: { initialValues: { switchField: true } },
  },
  args: {
    name: 'switchField',
    label: '🌙 Dark mode developer',
    thumbIcon: ({ isSelected, className }) => {
      return isSelected ? (
        <FaRocket className={className} />
      ) : (
        <FaFlask className={className} />
      );
    },
  },
};

export const WithStartAndEndContent: Story = {
  parameters: {
    formProps: { initialValues: { switchField: true } },
  },
  args: {
    name: 'switchField',
    label: '📧 Send emails at 3 AM',
    startContent: <FaPhone />,
    endContent: <FaEnvelope />,
  },
};

export const AllSizes: Story = {
  render: () => {
    return (
      <div className="flex flex-col gap-4">
        <Switch label="🐭 Microservices" name="smallSwitch" size="sm" />
        <Switch label="🏗️ Monolith" name="mediumSwitch" size="md" />
        <Switch label="🏢 Enterprise Monolith" name="largeSwitch" size="lg" />
      </div>
    );
  },
};
