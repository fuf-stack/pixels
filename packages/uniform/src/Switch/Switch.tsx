import type { TVClassName, TVProps } from '@fuf-stack/pixel-utils';

import { Switch as NextSwitch } from '@nextui-org/switch';

import { tv, variantsToClassNames } from '@fuf-stack/pixel-utils';

import { Controller } from '../Controller';
import { useFormContext, useInput } from '../hooks';
import { FieldCopyTestIdButton } from '../partials/FieldCopyTestIdButton';
import { FieldValidationError } from '../partials/FieldValidationError';

export const switchVariants = tv({
  slots: {
    base: '',
    endContent: '',
    errorMessage: 'ml-1 mt-1',
    // See NextUI styles for group-data condition, e.g.: https://github.com/nextui-org/nextui/blob/main/packages/core/theme/src/components/select.ts
    label:
      'text-sm text-foreground subpixel-antialiased group-data-[invalid=true]:!text-danger group-data-[required=true]:after:ml-0.5 group-data-[required=true]:after:text-danger group-data-[required=true]:after:content-["*"]',
    outerWrapper: 'place-content-center',
    startContent: '',
    thumb: '',
    thumbIcon: '',
    wrapper: '',
  },
});

type VariantProps = TVProps<typeof switchVariants>;
type ClassName = TVClassName<typeof switchVariants>;

export interface SwitchProps extends VariantProps {
  /** CSS class name */
  className?: ClassName;
  /** whether the select should be disabled */
  disabled?: boolean;
  /** component displayed next to the switch. */
  label?: React.ReactNode;
  /** name the field is registered under */
  name: string;
  /** HTML data-testid attribute used in e2e tests */
  testId?: string;
}

/**
 *  Switch component based on [NextUI Switch](https://nextui.org/docs/components/switch)
 */
const Switch = ({
  className = undefined,
  disabled = false,
  label: _label = undefined,
  name,
  testId: _testId = undefined,
}: SwitchProps) => {
  const { control, debugMode, getFieldState } = useFormContext();
  const { error, required, testId, invalid } = getFieldState(name, _testId);

  const { label, getInputProps, getErrorMessageProps } = useInput({
    errorMessage: JSON.stringify(error),
    isInvalid: invalid,
    isRequired: required,
    label: _label,
    labelPlacement: 'outside',
    placeholder: ' ',
  });

  const variants = switchVariants();
  const classNames = variantsToClassNames(variants, className, 'outerWrapper');

  const showTestIdCopyButton = debugMode === 'debug-testids';

  return (
    <Controller
      name={name}
      control={control}
      disabled={disabled}
      render={({
        field: { disabled: isDisabled, value, ref, onBlur, onChange },
      }) => {
        return (
          <div className={classNames.outerWrapper}>
            <NextSwitch
              aria-describedby={getInputProps()['aria-describedby']}
              classNames={classNames}
              // See NextUI styles for group-data condition (data-invalid), e.g.: https://github.com/nextui-org/nextui/blob/main/packages/components/select/src/use-select.ts
              data-invalid={invalid}
              data-required={required}
              data-testid={testId}
              isDisabled={isDisabled}
              isSelected={!!value}
              name={name}
              onBlur={onBlur}
              onValueChange={onChange}
              ref={ref}
              required={required}
              value={value}
            >
              {label}
              {showTestIdCopyButton && (
                <FieldCopyTestIdButton testId={testId} />
              )}
            </NextSwitch>
            {error && (
              <div className={classNames.errorMessage}>
                <div
                  /* eslint-disable-next-line react/jsx-props-no-spreading */
                  {...getErrorMessageProps()}
                >
                  <FieldValidationError error={error} />
                </div>
              </div>
            )}
          </div>
        );
      }}
    />
  );
};

export default Switch;
