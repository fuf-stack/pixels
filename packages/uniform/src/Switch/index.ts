import Switch from './Switch';

export type { SwitchProps } from './Switch';

export { Switch };

export default Switch;
