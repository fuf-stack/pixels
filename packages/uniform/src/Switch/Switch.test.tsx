import { describe } from 'vitest';

import storySnapshots from '@repo/storybook-config/story-snapshots';

import * as stories from './Switch.stories';

describe.skip('Story Snapshots', () => {
  storySnapshots(stories);
});
