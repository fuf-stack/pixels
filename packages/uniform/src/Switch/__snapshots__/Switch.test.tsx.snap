// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Story Snapshots > Default 1`] = `
<div>
  <div
    class="flex w-full flex-row justify-between gap-6"
  >
    <form
      class="flex-grow min-w-60"
      data-testid=""
    >
      <div
        class="place-content-center"
      >
        <label
          class="group relative max-w-fit inline-flex items-center justify-start cursor-pointer touch-none tap-highlight-transparent select-none"
          data-invalid="false"
          data-required="false"
          data-testid="switchfield"
        >
          <input
            aria-labelledby=":r0:"
            class="font-inherit text-[100%] leading-[1.15] m-0 p-0 overflow-visible box-border absolute top-0 w-full h-full opacity-[0.0001] z-[1] cursor-pointer disabled:cursor-default"
            name="switchField"
            role="switch"
            type="checkbox"
            value=""
          />
          <span
            aria-hidden="true"
            class="px-1 relative inline-flex items-center justify-start flex-shrink-0 overflow-hidden bg-default-200 rounded-full outline-none group-data-[focus-visible=true]:z-10 group-data-[focus-visible=true]:ring-2 group-data-[focus-visible=true]:ring-focus group-data-[focus-visible=true]:ring-offset-2 group-data-[focus-visible=true]:ring-offset-background group-data-[selected=true]:bg-primary group-data-[selected=true]:text-primary-foreground w-12 h-7 transition-background"
          >
            <span
              class="z-10 flex items-center justify-center bg-white shadow-small rounded-full origin-right pointer-events-none w-5 h-5 text-small group-data-[selected=true]:ms-5 transition-all group-data-[pressed=true]:w-6 group-data-[selected]:group-data-[pressed]:ml-4"
            />
          </span>
          <span
            class="relative select-none ms-2 text-sm text-foreground subpixel-antialiased group-data-[invalid=true]:!text-danger group-data-[required=true]:after:ml-0.5 group-data-[required=true]:after:text-danger group-data-[required=true]:after:content-["*"]"
            id=":r0:"
          />
        </label>
      </div>
      <div
        class="mt-4 flex justify-end"
      >
        <button
          class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent data-[pressed=true]:scale-[0.97] outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-4 min-w-20 h-10 text-small gap-2 rounded-medium [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-success data-[hover=true]:opacity-hover text-white"
          data-testid="form_submit_button"
          type="submit"
        >
          Submit
        </button>
      </div>
    </form>
  </div>
</div>
`;

exports[`Story Snapshots > Disabled 1`] = `
<div>
  <div
    class="flex w-full flex-row justify-between gap-6"
  >
    <form
      class="flex-grow min-w-60"
      data-testid=""
    >
      <div
        class="place-content-center"
      >
        <label
          class="group relative max-w-fit inline-flex items-center justify-start cursor-pointer touch-none tap-highlight-transparent select-none opacity-disabled pointer-events-none"
          data-disabled="true"
          data-invalid="false"
          data-required="false"
          data-testid="switchfield"
        >
          <input
            aria-labelledby=":r2:"
            class="font-inherit text-[100%] leading-[1.15] m-0 p-0 overflow-visible box-border absolute top-0 w-full h-full opacity-[0.0001] z-[1] cursor-pointer disabled:cursor-default"
            disabled=""
            name="switchField"
            role="switch"
            type="checkbox"
            value=""
          />
          <span
            aria-hidden="true"
            class="px-1 relative inline-flex items-center justify-start flex-shrink-0 overflow-hidden bg-default-200 rounded-full outline-none group-data-[focus-visible=true]:z-10 group-data-[focus-visible=true]:ring-2 group-data-[focus-visible=true]:ring-focus group-data-[focus-visible=true]:ring-offset-2 group-data-[focus-visible=true]:ring-offset-background group-data-[selected=true]:bg-primary group-data-[selected=true]:text-primary-foreground w-12 h-7 transition-background"
          >
            <span
              class="z-10 flex items-center justify-center bg-white shadow-small rounded-full origin-right pointer-events-none w-5 h-5 text-small group-data-[selected=true]:ms-5 transition-all group-data-[pressed=true]:w-6 group-data-[selected]:group-data-[pressed]:ml-4"
            />
          </span>
          <span
            class="relative select-none ms-2 text-sm text-foreground subpixel-antialiased group-data-[invalid=true]:!text-danger group-data-[required=true]:after:ml-0.5 group-data-[required=true]:after:text-danger group-data-[required=true]:after:content-["*"]"
            id=":r2:"
          />
        </label>
      </div>
      <div
        class="mt-4 flex justify-end"
      >
        <button
          class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent data-[pressed=true]:scale-[0.97] outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-4 min-w-20 h-10 text-small gap-2 rounded-medium [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-success data-[hover=true]:opacity-hover text-white"
          data-testid="form_submit_button"
          type="submit"
        >
          Submit
        </button>
      </div>
    </form>
  </div>
</div>
`;

exports[`Story Snapshots > Invalid 1`] = `
<div>
  <div
    class="flex w-full flex-row justify-between gap-6"
  >
    <form
      class="flex-grow min-w-60"
      data-testid=""
    >
      <div
        class="place-content-center"
      >
        <label
          aria-describedby="react-aria-react-useId-mock react-aria-react-useId-mock"
          class="group relative max-w-fit inline-flex items-center justify-start cursor-pointer touch-none tap-highlight-transparent select-none"
          data-focus="true"
          data-hover="true"
          data-invalid="true"
          data-required="true"
          data-selected="true"
          data-testid="switchfield"
          required=""
        >
          <input
            aria-labelledby=":r4:"
            class="font-inherit text-[100%] leading-[1.15] m-0 p-0 overflow-visible box-border absolute top-0 w-full h-full opacity-[0.0001] z-[1] cursor-pointer disabled:cursor-default"
            name="switchField"
            role="switch"
            type="checkbox"
            value="true"
          />
          <span
            aria-hidden="true"
            class="px-1 relative inline-flex items-center justify-start flex-shrink-0 overflow-hidden bg-default-200 rounded-full outline-none group-data-[focus-visible=true]:z-10 group-data-[focus-visible=true]:ring-2 group-data-[focus-visible=true]:ring-focus group-data-[focus-visible=true]:ring-offset-2 group-data-[focus-visible=true]:ring-offset-background group-data-[selected=true]:bg-primary group-data-[selected=true]:text-primary-foreground w-12 h-7 transition-background"
          >
            <span
              class="z-10 flex items-center justify-center bg-white shadow-small rounded-full origin-right pointer-events-none w-5 h-5 text-small group-data-[selected=true]:ms-5 transition-all group-data-[pressed=true]:w-6 group-data-[selected]:group-data-[pressed]:ml-4"
            />
          </span>
          <span
            class="relative select-none ms-2 text-sm text-foreground subpixel-antialiased group-data-[invalid=true]:!text-danger group-data-[required=true]:after:ml-0.5 group-data-[required=true]:after:text-danger group-data-[required=true]:after:content-["*"]"
            id=":r4:"
          >
            switch
          </span>
        </label>
        <div
          class="ml-1 mt-1"
        >
          <div
            class="text-tiny text-danger"
            data-slot="error-message"
            id="react-aria-react-useId-mock"
          >
            <ul
              aria-label="Error: I want to be switched off"
              data-testid="i_want_to_be_switched_off"
            >
              <li>
                <div>
                  I want to be switched off
                </div>
              </li>
            </ul>
          </div>
        </div>
      </div>
      <div
        class="mt-4 flex justify-end"
      >
        <button
          class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent data-[pressed=true]:scale-[0.97] outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-4 min-w-20 h-10 text-small gap-2 rounded-medium [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-success data-[hover=true]:opacity-hover text-white"
          data-testid="form_submit_button"
          type="submit"
        >
          Submit
        </button>
      </div>
    </form>
  </div>
</div>
`;

exports[`Story Snapshots > Required 1`] = `
<div>
  <div
    class="flex w-full flex-row justify-between gap-6"
  >
    <form
      class="flex-grow min-w-60"
      data-testid=""
    >
      <div
        class="place-content-center"
      >
        <label
          class="group relative max-w-fit inline-flex items-center justify-start cursor-pointer touch-none tap-highlight-transparent select-none"
          data-invalid="false"
          data-required="true"
          data-testid="switchfield"
          required=""
        >
          <input
            aria-labelledby=":r3:"
            class="font-inherit text-[100%] leading-[1.15] m-0 p-0 overflow-visible box-border absolute top-0 w-full h-full opacity-[0.0001] z-[1] cursor-pointer disabled:cursor-default"
            name="switchField"
            role="switch"
            type="checkbox"
            value=""
          />
          <span
            aria-hidden="true"
            class="px-1 relative inline-flex items-center justify-start flex-shrink-0 overflow-hidden bg-default-200 rounded-full outline-none group-data-[focus-visible=true]:z-10 group-data-[focus-visible=true]:ring-2 group-data-[focus-visible=true]:ring-focus group-data-[focus-visible=true]:ring-offset-2 group-data-[focus-visible=true]:ring-offset-background group-data-[selected=true]:bg-primary group-data-[selected=true]:text-primary-foreground w-12 h-7 transition-background"
          >
            <span
              class="z-10 flex items-center justify-center bg-white shadow-small rounded-full origin-right pointer-events-none w-5 h-5 text-small group-data-[selected=true]:ms-5 transition-all group-data-[pressed=true]:w-6 group-data-[selected]:group-data-[pressed]:ml-4"
            />
          </span>
          <span
            class="relative select-none ms-2 text-sm text-foreground subpixel-antialiased group-data-[invalid=true]:!text-danger group-data-[required=true]:after:ml-0.5 group-data-[required=true]:after:text-danger group-data-[required=true]:after:content-["*"]"
            id=":r3:"
          >
            switchField
          </span>
        </label>
      </div>
      <div
        class="mt-4 flex justify-end"
      >
        <button
          class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent data-[pressed=true]:scale-[0.97] outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-4 min-w-20 h-10 text-small gap-2 rounded-medium [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-success data-[hover=true]:opacity-hover text-white"
          data-testid="form_submit_button"
          type="submit"
        >
          Submit
        </button>
      </div>
    </form>
  </div>
</div>
`;

exports[`Story Snapshots > WithInitialValue 1`] = `
<div>
  <div
    class="flex w-full flex-row justify-between gap-6"
  >
    <form
      class="flex-grow min-w-60"
      data-testid=""
    >
      <div
        class="place-content-center"
      >
        <label
          class="group relative max-w-fit inline-flex items-center justify-start cursor-pointer touch-none tap-highlight-transparent select-none"
          data-invalid="false"
          data-required="false"
          data-selected="true"
          data-testid="switchfield"
        >
          <input
            aria-labelledby=":r1:"
            checked=""
            class="font-inherit text-[100%] leading-[1.15] m-0 p-0 overflow-visible box-border absolute top-0 w-full h-full opacity-[0.0001] z-[1] cursor-pointer disabled:cursor-default"
            name="switchField"
            role="switch"
            type="checkbox"
            value="true"
          />
          <span
            aria-hidden="true"
            class="px-1 relative inline-flex items-center justify-start flex-shrink-0 overflow-hidden bg-default-200 rounded-full outline-none group-data-[focus-visible=true]:z-10 group-data-[focus-visible=true]:ring-2 group-data-[focus-visible=true]:ring-focus group-data-[focus-visible=true]:ring-offset-2 group-data-[focus-visible=true]:ring-offset-background group-data-[selected=true]:bg-primary group-data-[selected=true]:text-primary-foreground w-12 h-7 transition-background"
          >
            <span
              class="z-10 flex items-center justify-center bg-white shadow-small rounded-full origin-right pointer-events-none w-5 h-5 text-small group-data-[selected=true]:ms-5 transition-all group-data-[pressed=true]:w-6 group-data-[selected]:group-data-[pressed]:ml-4"
            />
          </span>
          <span
            class="relative select-none ms-2 text-sm text-foreground subpixel-antialiased group-data-[invalid=true]:!text-danger group-data-[required=true]:after:ml-0.5 group-data-[required=true]:after:text-danger group-data-[required=true]:after:content-["*"]"
            id=":r1:"
          />
        </label>
      </div>
      <div
        class="mt-4 flex justify-end"
      >
        <button
          class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent data-[pressed=true]:scale-[0.97] outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-4 min-w-20 h-10 text-small gap-2 rounded-medium [&>svg]:max-w-[theme(spacing.8)] transition-transform-colors-opacity motion-reduce:transition-none bg-success data-[hover=true]:opacity-hover text-white"
          data-testid="form_submit_button"
          type="submit"
        >
          Submit
        </button>
      </div>
    </form>
  </div>
</div>
`;
